### PR TITLE
[JENKINS-53511] Improve discovery and readability of WebClient

### PR DESCRIPTION
**Pull request by Wadeck**
- downstream of https://github.com/jenkinsci/jenkins-test-harness/pull/109
- update of commons-io from 2.4 to 2.6 to be ok with JTH:2.42. It's fully compatible with 2.4 cf [upgrade to 2.5](https://commons.apache.org/proper/commons-io/upgradeto2_5.html) and [upgrade to 2.6](https://commons.apache.org/proper/commons-io/upgradeto2_6.html).

See [JENKINS-53511](https://issues.jenkins-ci.org/browse/JENKINS-53511).

### Proposed changelog entries

* Developer: Improve discovery and readability of WebClient most popular options.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by jeffret-b — APPROVED** [review]
This cleans things up nicely.

**Review by fcojfernandez — APPROVED** [review]

File: test/pom.xml
Comment (jglick) [inline test/pom.xml]: please add a `TODO` comment linking to upstream

File: test/src/test/java/hudson/TcpSlaveAgentListenerTest.java
Comment (jglick) [inline test/src/test/java/hudson/TcpSlaveAgentListenerTest.java]: can also use `assertFails`

File: test/pom.xml
Comment (jglick) [reply test/pom.xml]: Anyway I released 2.42.

Comment (oleg-nenashev) [reply test/pom.xml]: #3661 picks it up

**Comment by oleg-nenashev** [comment]
@Wadeck are you still interested in it? There are merge conflicts, and there will be more if it is not merged soon

**Comment by Wadeck** [comment]
For the information, the upgrade from `jenkins-test-harness:2.41.1` to `jenkins-test-harness:2.46` had 2 broken backward compabitility:
- `XmlPage` removed the method `getContent` that was deprecated, so I replace it call by `getWebResponse().getContentAsString()`, in [`ViewTest`](https://github.com/jenkinsci/jenkins/pull/3618/files#diff-aca66d2a0c1f0a5da6c22e96e923c069R251).
- `WebResponse` changed the signature of the method `getContentAsString(String)` by `getContentAsString(Charset)`, so I replaced `"UTF-8"` by `StandardCharsets.UTF_8` in [`ComputerTest`](https://github.com/jenkinsci/jenkins/pull/3618/files#diff-9d99070dbe9cb3d19630a2d630d4862eR110).

**Comment by oleg-nenashev** [comment]
> For the information, the upgrade from jenkins-test-harness:2.41.1 to jenkins-test-harness:2.46 had 2 broken backward compabitility issues:

It has broken even more. I tried updating from 2.41 to 2.42 in https://github.com/jenkinsci/jenkins/pull/3661 , but finally I gave up and released 2.41.1

**Comment by Wadeck** [comment]
> It has broken even more

Seems I have only discovered the top of the iceberg :p will try to correct them...

**Review by jvz — APPROVED** [review]
Nice cleanup!

File: test-pom/pom.xml
Comment (vilacides) [inline test-pom/pom.xml]: Are these versions commented on purpose?

Comment (Wadeck) [reply test-pom/pom.xml]: I am still correcting the tests, so it's WiP

Comment (Wadeck) [reply test-pom/pom.xml]: Ho actually I thought I pushed the 2.46 after the failing try with 2.42. Thx for the headsup!

**Comment by Wadeck** [comment]
Next step, try using system property to enforce the `withoutSpace` behavior to avoid troubles with the temp directories.

File: test-pom/pom.xml
Comment (batmat) [reply test-pom/pom.xml]: I think this bump should actually be reverted in not required for the scope of that PR. This seems like something that can (and if can, then should) be handled in a dedicated PR for clarity (and more speedy reviews because less code, good for everyone).

**Comment by batmat** [comment]
in https://github.com/jenkinsci/jenkins/pull/3618/commits/72bafd9d4f8dd72a084827cd26e0a2df173bbc29

> Re-put the 2.46 that is the target

Why is this actually the target?

**Comment by Wadeck** [comment]
As discussed with Baptiste, this PR will only focus on inclusion of the JTH version that provide the initial desired methods, not all the change with the space issue. So putting the JTH to 2.42 only.

Due to this update, I was forced to bump as well the `commons-io` lib to 2.6. It's fully compatible with 2.4 cf [upgrade to 2.5](https://commons.apache.org/proper/commons-io/upgradeto2_5.html) and [upgrade to 2.6](https://commons.apache.org/proper/commons-io/upgradeto2_6.html).

**Comment by jglick** [comment]
> The problem with the space is that `StreamResult#new` convert the file to URI and so, convert the space into `%20`, that is not recognized as a file.

Sounds like a legitimate bug, or what am I missing?

**Comment by Wadeck** [comment]
> Sounds like a legitimate bug, or what am I missing?

The constructor transforms the File into URI but that's the good behavior at this point. I *thought* it was the root cause but actually it's the interpretation of this URI that is badly done. That bad interpretation is only present in the Xalan version included in the JTH.

I tested JTH-2.42 with / without space, same behavior. Using `hpi:run`, (with JTH:2.42) it works with and without space.

The JTH:2.46 has trouble with space but not without. And if I run the `hpi:run` with JTH:2.46, it works with and without space.

Conclusion, it's only a problem in test and so, not a "production" bug.

-----

Thanks @thomasgl-orange for the so simple PR that solves my annoying issue!
Are you OK if we merge this PR and then you can adjust yours to just rebase it on the master after the merge? As it's that simple (removal of a dep) it should be easy. Otherwise I can integrate your change here, let me know what you prefer!

**Comment by thomasgl-orange** [comment]
> Are you OK if we merge this PR and then you can adjust yours to just rebase it on the master after the merge?

Sure, sounds good, once this one gets merged I will update mine.

**Comment by Wadeck** [comment]
I revoked the previous approval as I made changes since their creation

**Review by fcojfernandez — APPROVED** [review]
It still looks good to me!

**Review by jvz — APPROVED** [review]

File: test-pom/pom.xml
Comment (vilacides) [reply test-pom/pom.xml]: This has been marked as ready for merge but the code is still commented. Shouldn't we just remove it?

**Review by alecharp — CHANGES_REQUESTED** [review]
two comments to remove on `test/pom.xml` and it's good.

**Review by vilacides — APPROVED** [review]

**Review by alecharp — APPROVED** [review]
I don't understand why Github showed me outdated content in the review..

**Comment by Wadeck** [comment]
Seems that Isa got the same issue

**Comment by Wadeck** [comment]
Squash of the commits to separate clearly what was required for the JTH bump (due to breaking change in HTMLUnit new version) and what's for the discovery of the new features.

**Review by oleg-nenashev — APPROVED** [review]
Thanks @Wadeck ! Will merge once the tests pass.

**Comment by oleg-nenashev** [comment]
i doubt this is related, but...

```
Stacktrace
java.lang.AssertionError
at hudson.slaves.PingThreadTest.failedPingThreadResetsComputerChannel(PingThreadTest.java:90)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:566)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:552)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.lang.Thread.run(Thread.java:834)
```

**Review by jeffret-b — APPROVED** [review]

**Comment by batmat** [comment]
I strongly suspect Oleg was planning to close/reopen to retrigger the build.

**Comment by oleg-nenashev** [comment]
@batmat Definitely. Sorry for forgetting about the second step :(

**Comment by batmat** [comment]
No problem *at all* Oleg. It's happened to me many times :)

**Comment by Wadeck** [comment]
And now it's passing, thx @batmat / @oleg-nenashev

**Comment by oleg-nenashev** [comment]
That's why I do poke commits instead of close/reopen :(

**Comment by oleg-nenashev** [comment]
Will merge it once the build passes

**Pull request merged**
